lib/linked_list.rb

```diff
@@ -7,39 +7,83 @@ def initialize
     @head = nil
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1)
+  # Space complexity - O(1), thanks to the power of linked lists we are not making any new data structures, just one new node.
   def add_first(data)
-
+    node = Node.new(data, @head)
+    @head = node
+    return @head.data
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def get_first
-
+    return @head.data if @head
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n), where n is the length of the list
+  # Space complexity - O(1)
   def length
-    return 0
+    count = 0
+    if @head
+      next_node = @head.next
+      count += 1
+      while next_node != nil
+        count += 1
+        next_node = next_node.next
+      end
+    end
+    return count
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n)
+  # Space complexity - O(1)
   def add_last(data)
-
+    if @head
+      if @head.next
+        next_node = @head.next
+        until next_node.next == nil
+          next_node = next_node.next
+        end
+        next_node.next = Node.new(data)
+      else
+        @head.next = Node.new(data)
+      end
+    else
+      @head = Node.new(data)
+    end
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n)
+  # Space complexity - O(1)
   def get_last
-
+    if @head.next
```

Comment:
What if `head` is `nil`?

```suggestion
    return nil if @head.nil?
    if @head.next
```